server/coordinator/node_picker.go

```diff
@@ -4,51 +4,47 @@ package coordinator
 
 import (
 	"context"
-	"crypto/rand"
-	"math/big"
+	"strconv"
 	"time"
 
+	"github.com/CeresDB/ceresmeta/pkg/log"
 	"github.com/CeresDB/ceresmeta/server/cluster/metadata"
+	"github.com/CeresDB/ceresmeta/server/hash"
+	"github.com/CeresDB/ceresmeta/server/storage"
 	"github.com/pkg/errors"
+	"go.uber.org/zap"
 )
 
 type NodePicker interface {
-	PickNode(ctx context.Context, registerNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error)
+	PickNode(ctx context.Context, shardID storage.ShardID, registerNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error)
 }
 
-type RandomNodePicker struct{}
+type ConsistentHashNodePicker struct {
+	hashReplicas int
+}
 
-func NewRandomNodePicker() NodePicker {
-	return &RandomNodePicker{}
+func NewConsistentHashNodePicker(hashReplicas int) NodePicker {
+	return &ConsistentHashNodePicker{hashReplicas: hashReplicas}
 }
 
-func (p *RandomNodePicker) PickNode(_ context.Context, registeredNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error) {
-	now := time.Now().Unix()
-	onlineNodeLength := 0
-	for _, registeredNode := range registeredNodes {
-		if !registeredNode.IsExpired(now) {
-			onlineNodeLength++
+func (p *ConsistentHashNodePicker) PickNode(_ context.Context, shardID storage.ShardID, registerNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error) {
+	now := time.Now()
+
+	hashRing := hash.New(p.hashReplicas, nil)
+	nodeMapping := make(map[string]metadata.RegisteredNode, len(registerNodes))
```

Comment:
I guess there is no need to construct this map? With the hash result, we can find the node just by looping the slice without any memory allocation.